bindings/src/main/java/io/pravega/storage/hdfs/HDFSChunkStorage.java

```diff
@@ -0,0 +1,350 @@
+        try {
+            val status = fileSystem.getFileStatus(getFilePath(chunkName));
+            if (status.getPermission().getUserAction() == FsAction.READ) {
+                return ChunkHandle.readHandle(chunkName);
+            } else {
+                return ChunkHandle.writeHandle(chunkName);
+            }
+        } catch (IOException e) {
+            throw convertException(chunkName, "doOpenWrite", e);
+        }
+
+    }
+
+    @Override
+    protected int doRead(ChunkHandle handle, long fromOffset, int length, byte[] buffer, int bufferOffset)
+            throws ChunkStorageException, NullPointerException, IndexOutOfBoundsException {
+        ensureInitializedAndNotClosed();
+        try (FSDataInputStream stream = this.fileSystem.open(getFilePath(handle.getChunkName()))) {
+            stream.readFully(fromOffset, buffer, bufferOffset, length);
+        } catch (EOFException e) {
+            throw new IllegalArgumentException(String.format("Reading at offset (%d) which is beyond the current size of chunk.", fromOffset));
+        } catch (IOException e) {
+            throw convertException(handle.getChunkName(), "doRead", e);
+        }
+        return length;
+    }
+
+    @Override
+    protected int doWrite(ChunkHandle handle, long offset, int length, InputStream data) throws ChunkStorageException, IndexOutOfBoundsException {
+        ensureInitializedAndNotClosed();
+        try (FSDataOutputStream stream = this.fileSystem.append(getFilePath(handle.getChunkName()))) {
+            if (stream.getPos() != offset) {
+                // Looks like the filesystem changed from underneath us. This could be our bug, but it could be something else.
+                throw new IndexOutOfBoundsException();
+            }
+
+            if (length == 0) {
+                // Note: IOUtils.copyBytes with length == 0 will enter an infinite loop, hence the need for this check.
+                return 0;
+            }
+
+            // We need to be very careful with IOUtils.copyBytes. There are many overloads with very similar signatures.
+            // There is a difference between (InputStream, OutputStream, int, boolean) and (InputStream, OutputStream, long, boolean),
+            // in that the one with "int" uses the third arg as a buffer size, and the one with "long" uses it as the number
+            // of bytes to copy.
+            IOUtils.copyBytes(data, stream, (long) length, false);
+
+            stream.flush();
+        } catch (IOException e) {
+            throw convertException(handle.getChunkName(), "doWrite", e);
+        }
+        return length;
+    }
+
+    @Override
+    public int doConcat(ConcatArgument[] chunks) throws ChunkStorageException {
+        ensureInitializedAndNotClosed();
+        int length = 0;
+        try {
+            val sources = new Path[chunks.length - 1];
+            this.fileSystem.truncate(getFilePath(chunks[0].getName()), chunks[0].getLength());
```

Comment:
what happens in case of failure during the execution of this method ?
we are losing the contents of the file IMHO

Reply:
There is a very specific segment store fail over situation when actual chunk has more data than its length recorded in metadata. This can happen when zombie instance keeps writing to chunk even after it is no longer the owner of the segment. The new segment store instance will effectively "seal" that chunk by recording the most recent length and adding a fresh new chunk. In short data after recorded length is garbage that needs to be ignored anyway.
See https://github.com/pravega/pravega/wiki/PDP-34:-Simplified-Tier-2#segment-store-failover

HDFS concat does not have a concept of concatenating a partial  file. So removing unneeded end part.

Segment concat is called in only two cases 
- for cases where transaction segment is merged into the main segment. 
- de-fragmenting an existing segment by merging smaller chunks into bigger chunks.
In both the cases above there is no possibility of having garbage data/invalid data at the front of the chunks.